File: app/src/main/java/com/jakebarnby/filemanager/ui/adapters/SourceLogoutAdapter.java
package com.jakebarnby.filemanager.ui.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.jakebarnby.filemanager.R;
import com.jakebarnby.filemanager.sources.models.Source;
import com.jakebarnby.filemanager.util.Utils;

import java.util.List;

/**
 * Created by Jake on 8/6/2017.
 */

public class SourceLogoutAdapter extends RecyclerView.Adapter<SourceLogoutAdapter.LogoutViewHolder> {

    private List<Source> mSources;
    private LogoutListener mListener;

    public SourceLogoutAdapter(List<Source> sources, LogoutListener listener) {
        this.mSources = sources;
        this.mListener = listener;
    }

    @Override
    public LogoutViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_logout_list, parent, false);
        return new LogoutViewHolder(view);
    }

    @Override
    public void onBindViewHolder(LogoutViewHolder holder, int position) {
        holder.bindHolder(mSources, position);
    }

    @Override
    public int getItemCount() {
        return mSources.size();
    }

    class LogoutViewHolder extends RecyclerView.ViewHolder {

        private ImageView   mLogo;
        private TextView    mSourceName;
        private Button      mLogout;

        LogoutViewHolder(View itemView) {
            super(itemView);
            this.mLogo          = itemView.findViewById(R.id.img_logo_logout);
            this.mSourceName    = itemView.findViewById(R.id.txt_source_title);
            this.mLogout        = itemView.findViewById(R.id.btn_logout);
        }

        void bindHolder(List<Source> sources, int position) {
            Source source = sources.get(position);

            mLogo.setImageResource(Utils.resolveLogoId(source.getSourceName()));
            mSourceName.setText(String.format("%s", source.getSourceName()));
            mLogout.setText(source.isLoggedIn() ? R.string.logout : R.string.connect);
            mLogout.setOnClickListener((view -> {
                if (source.isLoggedIn()) {
                    source.logout(view.getContext());
                    sources.remove(source);
                    notifyDataSetChanged();

                    if (sources.size() == 0) {
                        mListener.onLastLogout();
                    }
                } else {
                    source.authenticateSource(mLogo.getContext());
                }
            }));
        }
    }

    @FunctionalInterface
    public interface LogoutListener {
        void onLastLogout();
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/ui/adapters/SourceLogoutAdapter.kt
package com.jakebarnby.filemanager.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.jakebarnby.filemanager.R
import com.jakebarnby.filemanager.sources.models.Source
import com.jakebarnby.filemanager.ui.adapters.SourceLogoutAdapter.LogoutViewHolder
import com.jakebarnby.filemanager.util.Utils

/**
 * Created by Jake on 8/6/2017.
 */
class SourceLogoutAdapter(
    private val sources: MutableList<Source>,
    private val listener: LogoutListener
) : RecyclerView.Adapter<LogoutViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LogoutViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.view_logout_list, parent, false)
        return LogoutViewHolder(view)
    }

    override fun onBindViewHolder(holder: LogoutViewHolder, position: Int) {
        holder.bindHolder(sources, position)
    }

    override fun getItemCount(): Int {
        return sources.size
    }

    inner class LogoutViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val logo: ImageView = itemView.findViewById(R.id.img_logo_logout)
        private val sourceName: TextView = itemView.findViewById(R.id.txt_source_title)
        private val logout: Button = itemView.findViewById(R.id.btn_logout)

        fun bindHolder(sources: MutableList<Source>, position: Int) {

            val source = sources[position]

            logo.setImageResource(Utils.resolveLogoId(source.sourceName))
            sourceName.text = String.format("%s", source.sourceName)
            logout.setText(if (source.isLoggedIn) R.string.logout else R.string.connect)
            logout.setOnClickListener { view: View ->
                if (!source.isLoggedIn) {
                    source.authenticate(logo.context)
                    return@setOnClickListener
                }

                source.logout(view.context)
                sources.remove(source)
                notifyDataSetChanged()
                if (sources.size == 0) {
                    listener.onLastLogout()
                }
            }
        }

    }

    @FunctionalInterface
    interface LogoutListener {
        fun onLastLogout()
    }

}